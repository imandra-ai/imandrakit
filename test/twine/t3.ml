[@@@warning "-60"]

module M1 = struct
  Format.printf "without hashcons@."

  type t0 = {
    x: int; [@key "X"]
    y: bool; [@key "Y"]
  }
  [@@deriving twine, show]

  type t1 = { foos: t0 list } [@@deriving twine, show]

  let myt1 =
    let x1 = { x = 0; y = false } in
    let x2 = { x = 2; y = true } in
    { foos = [ x1; x1; x2; x1; x2 ] }
  ;;

  Format.printf "myt1: %a@." pp_t1 myt1

  let pack1 = Imandrakit_twine.Encode.to_string t1_to_twine myt1
  let () = Format.printf "pack1:@.%s@." (Hex.hexdump_s @@ Hex.of_string pack1)
  let () = Format.printf "len1: %d@." @@ String.length pack1
end

module M2 = struct
  Format.printf "with caching@."

  type t0 = {
    x: int; [@key "x0"]
    y: bool; [@key "y"]
  }
  [@@deriving twine, show, eq]

  module T0 = struct
    type t = t0 [@@deriving eq]

    let hash = Hashtbl.hash
  end

  type t1 = { foos: t0 list } [@@deriving twine, show]

  let () = Imandrakit_twine.Encode.add_cache (module T0) t0_to_twine_ref

  let myt1 =
    let x1 = { x = 0; y = false } in
    let x2 = { x = 2; y = true } in
    { foos = [ x1; x1; x2; x1; x2 ] }
  ;;

  Format.printf "myt1': %a@." pp_t1 myt1

  let pack1 = Imandrakit_twine.Encode.to_string t1_to_twine myt1
  let () = Format.printf "pack1:@.%s@." (Hex.hexdump_s @@ Hex.of_string pack1)
  let () = Format.printf "len1: %d@." (String.length pack1)
end
