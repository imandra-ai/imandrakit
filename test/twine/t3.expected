without hashcons
myt1: { T3.M1.foos =
        [{ T3.M1.x = 0; y = false }; { T3.M1.x = 0; y = false };
          { T3.M1.x = 2; y = true }; { T3.M1.x = 0; y = false };
          { T3.M1.x = 2; y = true }]
        }
pack1:
00000000: 6210 0062 1000 6212 0162 1000 6212 0165  b..b..b..b..b..e
00000001: ff00 fefc faf8 61f7 01                   ......a..

dump(t1):
[0x0]: [0, false] (len=2)
[0x3]: [0, false] (len=2)
[0x6]: [2, true] (len=2)
[0x9]: [0, false] (len=2)
[0xc]: [2, true] (len=2)
[0xf]: [@0x0, @0x3, @0x6, @0x9, @0xc] (len=5)
[0x16]: [@0xf] (len=1)

len1: 25
with caching
myt1': { T3.M2.foos =
         [{ T3.M2.x = 0; y = false }; { T3.M2.x = 0; y = false };
           { T3.M2.x = 2; y = true }; { T3.M2.x = 0; y = false };
           { T3.M2.x = 2; y = true }]
         }
pack1:
00000000: 6210 0062 1201 65f6 f7f5 f9f7 61f6 01    b..b..e.....a..

dump(t1'):
[0x0]: [0, false] (len=2)
[0x3]: [2, true] (len=2)
[0x6]: [@0x0, @0x0, @0x3, @0x0, @0x3] (len=5)
[0xc]: [@0x6] (len=1)

len1: 15
